Add production-ready gcloud token revocation script

happygoluckycodeeditor opened the pull request:
## Overview
Adds a production-ready version of the gcloud token revocation script for security teams to use during actual incidents.

## What's Added
- New file: `revoke-gcloud-production.gs` - Production script with safety features
- Updated: `README.md` - Concise documentation for both scripts

## Key Features
- Dry run mode for safe testing
-  Admin verification and testing functions  
-  Detailed logging and error handling
-  Built-in safety limits
-  Execution statistics

## Testing
Tested with dry run mode and verified all safety features and functions, and they work as expected.

## Breaking Changes
None - purely additive. Original script unchanged.

よろしくお願いいたします。

happygoluckycodeeditor committed:
Added production-ready gcloud token revocation script

- Add comprehensive production script with safety features
- Include dry run mode for safe testing
- Add detailed logging and error handling
- Update README with concise usage instructions

happygoluckycodeeditor committed:
Update README.md

happygoluckycodeeditor committed:
Delete .DS_Store

eeaton commented:
I noticed that the automated checks for Lint in Github Actions are failing with a 403 error, that must be a configuration issue on the base repo. I'll do a quick fix on that issue to unblock the automated tests running here.

eeaton committed:
Merge branch 'main' into main

eeaton commented:
@happygoluckycodeeditor the issues with the linter action should be resolved now, go ahead and update your branch with the latest version of Main.

eeaton committed:
Merge branch 'main' into main

eeaton commented:
Actually nevermind, I was able to update your branch and trigger the tests myself.

It seems the lint test for MARKDOWN_PRETTIER is failing, but with a vague error:

> MARKDOWN_PRETTIER
  2025-07-03 09:41:38 [INFO]   Linting MARKDOWN_PRETTIER items...
  Error: -03 09:41:38 [ERROR]   Found errors when linting MARKDOWN_PRETTIER. Exit code: 1.
  2025-07-03 09:41:39 [INFO]   Command output for MARKDOWN_PRETTIER:
  Checking formatting...
  2025-07-03 09:41:39 [INFO]   Stderr contents for MARKDOWN_PRETTIER:
  [warn] README.md
  [warn] Code style issues found in the above file. Run Prettier with --write to fix.

Try running [prettier](https://prettier.io/docs/install) against your README.md file and updating the pr.

happygoluckycodeeditor commented:
Let me run Prettier and get back to you! Thank you so much~!

eeaton commented on revoke-gcloud-production.gs:
```diff
@@ -0,0 +1,293 @@
+    // Try to get information about the current user
+    const userInfo = AdminDirectory.Users.get(currentUser);
+    Logger.log('✓ Can access user info for current user');
+    Logger.log('User info: %s', JSON.stringify(userInfo, null, 2));
```
This writes a lot of sensitive info about the admin to the logs, that isn't needed for this use case.

I recommend to remove L114, because L113 is sufficient for the purposes of this check.

happygoluckycodeeditor replied:
Removed L114

eeaton commented on revoke-gcloud-production.gs:
```diff
@@ -0,0 +1,293 @@
+    Logger.log('Response type: %s', typeof response);
+    
+    if (response.users) {
+      Logger.log('✓ Found %s users', response.users.length);
```
For this line and others that write numbers to the logs, the output is treated as floats, not integers ("5.0" instead of "5"), which makes it a bit harder to read.

I think you can fix this by wrapping the numbers in parseInt, like
`Logger.log('✓ Found %s users', parseInt(response.users.length));`

happygoluckycodeeditor replied:
Updated this line

eeaton replied:
I still see the output as "5.0" with floats, but after some experimentation, I think that might be an unavoidable behavior of Logger.log, and my recommended change didn't do anything. I tried a few different methods to change the format of users.length, but it seems that Logger.log() will always cast it with a decimal. (Compare to console.log, which gives the expected behavior of "5".) 

In general, using Logger.log is preferred over Console.log because it has better integration with GCP logging, but Console.log might be a workaround to get the preferred output. However, the format is such a minor cosmetic nitpick that I don't think it's worthwhile to revise each line against recommended practice. If there's some other formatting option for Logger.log I've missed, please let me know.

As the simplest fix, how about you undo this change (use just `response.users.length` instead of parseInt(response.users.length`) and then I'll approve it. I can live with the slightly odd number formatting in debug logs.

happygoluckycodeeditor replied:
I tried checking with other ways but I didn't find a good solution unfortunately. I have reverted back to the previous `response.users.length`. Have a look!
Thank you so much for taking the time to check the commit.

eeaton commented on revoke-gcloud-production.gs:
```diff
@@ -0,0 +1,293 @@
+        
+        // Check each token to find gcloud CLI tokens
+        for (const token of tokens) {
+          Logger.log('  Token client ID: %s', token.clientId);
```
Here's another place where I think too much is getting written to logs.
The tokenId isn't necessarily sensitive, and the Admin can already view it in other places on the Admin console, but I think writing all the other token-Ids to logs isn't needed for this use case and adds the possibility that info in logs gets exposed to someone who shouldn't have access.

happygoluckycodeeditor replied:
I improved the code a bit! よろしくお願いします！

eeaton commented on revoke-gcloud-production.gs:
```diff
@@ -0,0 +1,293 @@
+        const tokens = tokensResponse.items;
+        
+        // Log token information for this user
+        if (!tokens || tokens.length === 0) {
```
This section can be simplified by filtering on the array upfront, which would remove the need for some later arguments like `userhasGcloudToken` flag and `gcloudTokensCount`, and the intermediate steps of printing out all the token clientIDs including stuff not related to gcloud. The logic "if the tokens array exists and is non-0 and has any clientID values that match GCLOUD_CLIENT_ID" can be expressed in a single line with something like:

~~~

// Filter for gcloud CLI tokens first
const gcloudTokens = tokens ? tokens.filter(token => token.clientId === GCLOUD_CLIENT_ID) : [];

if (gcloudTokens.length === 0) {
  Logger.log('  No gcloud CLI OAuth tokens found for user: %s', user.primaryEmail);
  return; 
}

//then do the remaining actions with gcloudTokens, which only has the tokens we're interested in

happygoluckycodeeditor replied:
I have Updated the function!

eeaton reviewed (changes requested):
Functionality is overall good, but a few small fixes in comments:

happygoluckycodeeditor committed:
address pr feedback: simplify token filtering and logging

happygoluckycodeeditor commented:
Hi eeaton, so sorry for the late reply. I was heavily occupied with my company's internal project, I didn't notice the message.
All requested changes have been addressed and the code passes local lint checks (I ran it locally). Give it a review!
よろしくお願いいたします！

happygoluckycodeeditor committed:
address PR feedback: revert back to response.users.length

happygoluckycodeeditor commented:
When I ran super-linter locally it did not show any error.

I will run prettier against the README file and do it again. 
I think its showing errors on Mardown files, let me do that quick update. Hope this works.

eeaton committed:
Update README.md

I identified the lint issues that were failing the super-linter github action. Here's a fix locally

eeaton commented:
I recreated super-linter locally and ran it in fix mode, it was able to pass the checks after tweaking some minor whitespace issues so I updated those directly to your readme file. I'm running the GitHub Action one more time to verify

eeaton commented:
I'll approve and merge it now, thanks for the contribution!

happygoluckycodeeditor commented:
Thank you so much! It has been great working with you!
   I learnt a lot! Hope to work with you again in the future~!
